chore: release notes for 4.10.0 (#4857)
leanprover/lean4: RELEASES.md

  * [b056a0](https://github.com/leanprover/lean4/commit/b056a0b395bb728512a3f3e83bf9a093059d4301) adapts kernel interruption to the new cancellation system.
* **Cleanup:** [#4380](https://github.com/leanprover/lean4/pull/4380), [#4431](https://github.com/leanprover/lean4/pull/4431), [#4494](https://github.com/leanprover/lean4/pull/4494), [e8f768](https://github.com/leanprover/lean4/commit/e8f768f9fd8cefc758533bc76e3a12b398ed4a39), [de2690](https://github.com/leanprover/lean4/commit/de269060d17a581ed87f40378dbec74032633b27), [d3a756](https://github.com/leanprover/lean4/commit/d3a7569c97123d022828106468d54e9224ed8207), [#4404](https://github.com/leanprover/lean4/pull/4404), [#4537](https://github.com/leanprover/lean4/pull/4537).

### Compiler, runtime, and FFI

* [d85d3d](https://github.com/leanprover/lean4/commit/d85d3d5f3a09ff95b2ee47c6f89ef50b7e339126) fixes criterion for tail-calls in ownership calculation.
* [#3963](https://github.com/leanprover/lean4/pull/3963) adds validation of UTF-8 at the C++-to-Lean boundary in the runtime.
* [#4512](https://github.com/leanprover/lean4/pull/4512) fixes missing unboxing in interpreter when loading initialized value.

### Lake

> * [#4384](https://github.com/leanprover/lean4/pull/4384) deprecates `inputFile` and replaces it with `inputBinFile` and `inputTextFile` so that line ending normalization only affects text files.

Reviewer: ```suggestion
* [#4384](https://github.com/leanprover/lean4/pull/4384) deprecates `inputFile` and replaces it with `inputBinFile` and `inputTextFile`. Unlike `inputBinFile` (and `inputFile`),  `inputTextFile` normalizes line endings, which helps ensure text file traces are platform-independent. 
```

The current wording seems to imply the deprecated `inputFile` performed line ending normalization.

* [#4371](https://github.com/leanprover/lean4/pull/4371) simplifies dependency resolution code.

> * [#4446](https://github.com/leanprover/lean4/pull/4446) switches to use `src/lake/lakefile.toml` to avoid needing to load a version of Lake to build Lake.

Reviewer: I suggest moving this to the DevOps section.

* [#4439](https://github.com/leanprover/lean4/pull/4439) touches up the Lake configuration DSL and makes other improvements:
  string literals can now be used instead of identifiers for names,
  avoids using French quotes in `lake new` and `lake init` templates,
  changes the `exe` template to use `Main` for the main module,
  improves the `math` template error if `lean-toolchain` fails to download,
  and downgrades unknown configuration fields from an error to a warning to improve cross-version compatibility.
* [#4496](https://github.com/leanprover/lean4/pull/4496) tweaks `require` syntax and updates docs. Now `require` in TOML for a package name such as `doc-gen4` does not need French quotes.
* [#4485](https://github.com/leanprover/lean4/pull/4485) fixes a bug where package versions in indirect dependencies would take precedence over direct dependencies.

> * [#4518](https://github.com/leanprover/lean4/pull/4518) makes trace reading more robust. Lake now rebuilds if trace files are invalid or unreadable and is backwards compatible with previous pure numeric traces.

Reviewer: ```suggestion
```

This was backported to `v4.9.0`.

Author: Is this already in the release notes for 4.9.0? (PR numbers are hard to cross-reference if the backported PR doesn't mention which PR it comes from.)

Reviewer: @kmill Unfortunately, I do not think it was added to the v4.9.0 release notes. However, the PR and its backport do link to each other (the backport is #4520).

Author: I see, I need to make sure to copy the backported PR number into the release notes for the future.

> * [#4477](https://github.com/leanprover/lean4/pull/4477) exposes the bundled C compiler's compiler flags, which is needed to skip the use of `leanc` in Lake.

Reviewer: ```suggestion
* [#4477](https://github.com/leanprover/lean4/pull/4477) exposes the bundled C compiler's compiler flags
```

This is not really a Lake change (yet), so it may be best to move to a different category.

* [#4478](https://github.com/leanprover/lean4/pull/4478) fixes a bug where Lake incorrectly included the module dynamic library in a platform-independent trace.
* [#4529](https://github.com/leanprover/lean4/pull/4529) fixes some issues with bad import errors.
  A bad import in an executable no longer prevents the executable's root
  module from being built. This also fixes a problem where the location
  of a transitive bad import would not been shown.
  The root module of the executable now respects `nativeFacets`.
* [#4564](https://github.com/leanprover/lean4/pull/4564) fixes a bug where non-identifier script names could not be entered on the CLI without French quotes.

> * [#4566](https://github.com/leanprover/lean4/pull/4566) addresses a few issues with precompiled libraries.

Reviewer: ```suggestion
* [#4566](https://github.com/leanprover/lean4/pull/4566) addresses a few issues with precompiled libraries.
  * Fixes a bug where Lake would always precompile the package of a module.
  * If a module is precompiled, it now precompiles its imports. Previously, it would only do this if imported.
```

These changes seem significant enough to mention here,

> * [#4495](https://github.com/leanprover/lean4/pull/4495)
>   adds a new type of `require` that fetches package metadata from a
>   registry API endpoint (e.g. Reservoir) and then clones a Git package
>   using the information provided. To require such a dependency, the new
>   syntax is:
> 
>   ```lean
>   require <scope> / <pkg-name> [@ "git#<rev>"]
>   -- Example:
>   require "leanprover" / "doc-gen4"
>   ```

Reviewer: ````suggestion
* [#4495](https://github.com/leanprover/lean4/pull/4495), [#4692](https://github.com/leanprover/lean4/pull/4692), [#4849](https://github.com/leanprover/lean4/pull/4849)
  add a new type of `require` that fetches package metadata from a
  registry API endpoint (e.g. Reservoir) and then clones a Git package
  using the information provided. To require such a dependency, the new
  syntax is:

  ```lean
  require <scope> / <pkg-name> [@ git <rev>]
  -- Example:
  require "leanprover" / "doc-gen4"
  ```
````

This merges the touchups below into this as a single feature.

  Or in TOML:
  ```toml
  [[require]]
  name = "<pkg-name>"
  scope = "<scope>"
  rev = "<rev>"
  ```

  Unlike with Git dependencies, Lake can make use of the richer
  information provided by the registry to determine the default branch of
  the package. This means for repositories of packages like `doc-gen4`
  which have a default branch that is not `master`, Lake will now use said
  default branch (e.g., in `doc-gen4`'s case, `main`).

  Lake also supports configuring the registry endpoint via an environment
  variable: `RESERVIOR_API_URL`. Thus, any server providing a similar
  interface to Reservoir can be used as the registry. Further
  configuration options paralleling those of Cargo's [Alternative Registries](https://doc.rust-lang.org/cargo/reference/registries.html)
  and [Source Replacement](https://doc.rust-lang.org/cargo/reference/source-replacement.html)
  will come in the future.

> * [#4692](https://github.com/leanprover/lean4/pull/4692) and [#4849](https://github.com/leanprover/lean4/pull/4849) add and document `require @ git`.
>   This is syntactic sugar specifying a git revision as a dependency version in a `require` command. For example:
>   ```lean
>   require "leanprover-community" / "proofwidgets" @ git "v0.0.39"
>   ```

Reviewer: ```suggestion
```

Merged into the above.
